README

Summary
-------

The automatic addressbook plugin collect each address you send an email to and records it in an address book, making it available for later use or auto-completion.

report bugs, ideas and patches at : https://code.crapouillou.net/projects/roundcube-plugins


Contributing
------------

Feel free to report issues, suggest features or to submit patches on the issue tracker.

You can get the code from svn at:

    svn co https://crapouillou.net/svn/rc_plugins/automatic_addressbook





Changelog
v0.5 :
* able to work wit rouncube 0.9 to 1.3, including the 1.2 and 1.X releases, deprecating the rcmail usage

v0.4 :
* Refresh contact list when editing a contact in collected addressbook and having on_edit_move_to_default set to true (#68)
* Refresh contact list when copying a contact from collected addressbook and having on_edit_move_to_default set to true (Fix #69)
* Added polish translation. Thanks to Marcin Dorota (#72)
* Added lithuanian translation. Thanks to Kestutis M (#73)

v0.3.1 :
* Fix a bug that prevents RC < 0.8 to send mail with plugin activated (#67)

v0.3 :
* Added nl_NL translation (thanks to Daniel)
* Updated german translation (thanks to Dennis Riehle)
* Added zh_CN translation (thaks to FDF)

v0.3-rc1 :
* Populated documentation files (README and INSTALL) to have documentation available when installing offline
* Added Changelog File
* Fixed compatibility with other addressbook_plugins (#46)
* Fixed Add contact button disabled in default abook after viewing collected abook (#58).
  Side effect is that you can now add contact in collected abook.
* Disabled contact groups in collected abook
* Added calls to exec_hook when creating or deleting a contact to improve compatibility with other plugins (thanks to Venia Legendi for submitting patch) (#56)
* Added user setting to use automatic abook for autocompletion or not. (#49)
* Updated database statements in SQL to reflect changes from RC 0.5
* Now call decode_address_list function from rcube_mime as it is deprecated from rcube_imap (#51)
* Fixed on_edit_move_to_default not working. Fixes #47. Also Fixes #25
* Defined plugin task list as advised in plugin writing documentation (maybe will improve performance ?)
* Added pear package.xml file. This way, the plugin is referenced in the about dialog of interface
* added vietnamese translation found on internet

v0.2 and before :
* previous releases 


install/index.php
<?php
define('INSTALL_PATH', dirname($_SERVER['SCRIPT_FILENAME']).'/../../../');
require_once("program/include/iniset.php");


PEAR::setErrorHandling(PEAR_ERROR_DIE);

function is_eol($line)
{
    if ($engine != 'mssql') {
        return preg_match('/;$/', trim($line));
    } else {
        return preg_match('/(^|.*\s*)GO$/', trim($line));
    }
}


function init_db($DB, $engine, $sql_dir)
{
    // read schema file
    $fname = $sql_dir.'/'.$engine.'.initial.sql';
    if ($lines = @file($fname, FILE_SKIP_EMPTY_LINES)) {
      $buff = '';
      foreach ($lines as $i => $line) {
        if (preg_match('/^--/', $line))
          continue;
        $buff .= $line . "\n";
        if (is_eol($line)){
          $res = $DB->query($buff);
	  echo "\n";
//	  if(PEAR::isError($res)) {
//       die("SQL error : " . $res->getMessage());
//	  }
          $buff = '';
        }
      }
      return true;
    }
    else {
      echo "<strong>$engine backend not supported. Only sqlite is supported at the moment)... contributions are welcome !</strong>\n";
      return false;
    }
}

$sql_dir = getcwd();
chdir(INSTALL_PATH); //To avoid issues with relative sqlite paths

$rcmail = rcmail::get_instance();
$parts = split('://', $rcmail->config->get("db_dsnw"));
$engine = $parts[0];

$init_res = init_db($rcmail->db, $engine, $sql_dir);

if ($init_res) {
  $res = "Ok<br />\n Automatic addressbook plugin is installed !";
  
} else {
  $res = 'Failled';
}
echo '<p>Building table (engine : '.$engine.')...<strong>'.$res.'</strong></p>';

?>


generatepackage.php
#!/usr/bin/php
<?php

require_once('PEAR/PackageFileManager2.php');
PEAR::setErrorHandling(PEAR_ERROR_DIE);

$options = array(
  'filelistgenerator' => 'svn', // this copy of our source code is a CVS checkout
  'simpleoutput' => true,
  'baseinstalldir' => '/', // The PEAR directory install location
  'packagedirectory' => dirname('___FILE___'), // We’ve put this file in the root source code dir
  'clearcontents' => true, // dump any old package.xml content (set false to append release)
  // no bundling of cvs/svn files or this generator file
  'ignore' => array('generate_package_xml.php', '.svn', '.cvs*'),
  'dir_roles' => array( // set up roles for some directories; the default is php
    'config' => 'cfg',
    'localization' => 'data',
    'SQL' => 'data',
    ),
);

$packagefile="package.xml";
// Oddly enough, this is a PHP source code package…

$packagexml = &PEAR_PackageFileManager2::importOptions($packagefile, $options);
$packagexml->setPackageType('php');

// Package name, summary and longer description

$packagexml->setPackage('automatic_addressbook');
$packagexml->setSummary('The automatic addressbook plugin collect each address you send an email to and records it in an address book, making it available for later use or auto-completion.');
$packagexml->setDescription("   * Simple plugin to register to \"collect\" all recipients of sent mail
   * to a dedicated address book (usefull for autocompleting email you
   * already used). User can choose in preferences (compose group) to
   * enable or disable the feature of this plugin.
   * Aims to reproduce the similar features of thunderbird or gmail.");

// The channel where this package is hosted. Since we’re installing from a local
// downloaded file rather than a channel we’ll pretend it’s from PEAR.

$packagexml->setChannel('pear.roundcube.net');

// Add some release notes!

$notes = 'v0.3 :
* Populated documentation files (README and INSTALL) to have documentation available when installing offline
* Added Changelog File
* Fixed compatibility with other addressbook_plugins (#46)
* Fixed Add contact button disabled in default abook after viewing collected abook (#58).
  Side effect is that you can now add contact in collected abook.
* Disabled contact groups in collected abook
* Added calls to exec_hook when creating or deleting a contact to improve compatibility with other plugins (thanks to Venia Legendi for submitting patch) (#56)
* Updated database statements in SQL to reflect changes from RC 0.5
* Now call decode_address_list function from rcube_mime as it is deprecated from rcube_imap (#51)
* Fixed on_edit_move_to_default not working. Fixes #47. Also Fixes #25
* Defined plugin task list as advised in plugin writing documentation (maybe will improve performance ?)

v0.2 and before :
* previous releases';
$packagexml->setNotes($notes);

// Add any known dependencies such as PHP version, extensions, PEAR installer

$packagexml->setPhpDep('5.1.4');
$packagexml->setPearinstallerDep('1.4.0');
$packagexml->addPackageDepWithChannel('required', 'PEAR', 'pear.php.net', '1.4.0');

// Other info, like the Lead Developers. license, version details and stability type

$packagexml->addMaintainer('lead', 'sblaisot', 'Sebastien Blaisot', 'sebastien@blaisot.org');
$packagexml->setLicense('GNU GPLv3+', 'http://www.gnu.org/licenses/gpl-3.0.html');
$packagexml->setAPIVersion('0.3');
$packagexml->setReleaseVersion('0.3b1');
$packagexml->setReleaseStability('beta');
$packagexml->setAPIStability('stable');

// Add this as a release, and generate XML content

$packagexml->addRelease();
$packagexml->generateContents();

// Pass a “make” flag from the command line or browser address to actually write
// package.xml to disk, otherwise just debug it for any errors

if (isset($_GET['make']) || (isset($_SERVER['argv']) && @$_SERVER['argv'][1] == 'make')) {
  $packagexml->writePackageFile();
} else {
  $packagexml->debugPackageFile();
}
?>


automatic_addressbook.php
<?php

/* vim: set expandtab tabstop=4 shiftwidth=4 softtabstop=4: */

/**
 * Automatic address book
 *
 *
 * Simple plugin to register to "collect" all recipients of sent mail
 * to a dedicated address book (usefull for autocompleting email you
 * already used). User can choose in preferences (compose group) to
 * enable or disable the feature of this plugin.
 * Aims to reproduce the similar features of thunderbird or gmail.
 *
 * @version 0.4
 * @author Jocelyn Delalande (slightly modified by Roland 'rosali' Liebl)
 * @author Sebastien Blaisot <sebastien@blaisot.org>
 * @website https://github.com/sblaisot/automatic_addressbook
 * @licence http://www.gnu.org/licenses/gpl-3.0.html GNU GPLv3+
 *
 */

 /*
 * Skeletton based on "example_addressbook" plugin.
 * Contact adding code inspired by addcontact.inc by Thomas Bruederli
 */

class automatic_addressbook extends rcube_plugin
{
    public $task = 'mail|addressbook|settings';
    private $abook_id = 'collected';
  
    /**
     * Initialize plugin
     */
    public function init()
    {
        $this->add_hook('addressbooks_list', array($this, 'address_sources'));
        $this->add_hook('addressbook_get', array($this, 'get_address_book'));
        $this->add_hook('message_sent', array($this, 'register_recipients'));
        $this->add_hook('preferences_list', array($this, 'settings_table'));
        $this->add_hook('preferences_save', array($this, 'save_prefs'));
        $this->add_hook('contact_update', array($this, 'handle_doubles'));
        $this->add_hook('contact_create', array($this, 'handle_doubles'));
       
        $this->add_texts('localization/', false);
        $this->load_config('config/config.inc.php.dist');
        if (file_exists("./plugins/automatic_addressbook/config/config.inc.php")) {
            $this->load_config('config/config.inc.php');
        }

        // Adds an address-book category in rc <= 0.5, retro-compatibility code, 
        // not needed for rc 0.6
        $this->add_hook('preferences_sections_list', array($this, 'add_addressbook_category'));


        // use this address book for autocompletion queries
        $config = rcube::get_instance()->config;
        $sources = $config->get('autocomplete_addressbooks', array('sql'));
        
        if (!in_array($this->abook_id, $sources) &&
                      $config->get('use_auto_abook', true) &&
                      $config->get('use_auto_abook_for_completion', true)) {
            $sources[] = $this->abook_id;
            $config->set('autocomplete_addressbooks', $sources);
        }
    }
    
    /**
     * Register automatic_addressbook as address source
     *
     * @param  array $p  Hash array with list of available address books
     * @return array $p  Hash array with list of available address books
     */
    public function address_sources($p)
    {
        $rcube = rcube::get_instance();
        if ($rcube->config->get('use_auto_abook', true)) {
            $p['sources'][$this->abook_id] = 
                array('id' => $this->abook_id,
                      'name' => rcube_utils::rep_specialchars_output($this->gettext('automaticallycollected')),
                      'readonly' => FALSE, 'groups' => false);
        }

        return $p;
    }
  
    /**
     * Requests automatic_addressbook instance
     *
     * @param  array $p  Hash array containing the id of the requested abook
     * @return array $p  Hash array containing instance of the requested abook
     */
    public function get_address_book($p)
    {
        $rcube = rcube::get_instance();
        if (($p['id'] === $this->abook_id) && $rcube->config->get('use_auto_abook', true)) {
            require_once dirname(__FILE__) . '/automatic_addressbook_backend.php';
            $p['instance'] = new automatic_addressbook_backend($rcube->db, $rcube->user->ID);
            $p['instance']->groups = false;
        }
        return $p;
    }

    /**
     * Collect the email address of a just-sent email recipients into
     * the automatic addressbook (if it's not already in another
     * addressbook). 
     *
     * @param array $p Hash array containing header and body of sent mail
     * @return nothing
     */
    public function register_recipients($p)
    {
        $rcube = rcube::get_instance();
    
        if (!$rcube->config->get('use_auto_abook', true)) {
            return;
        }
    
        $headers = $p['headers'];

        if (!class_exists('rcube_mime')) { // RC < 0.8 compatibility code
            $IMAP = new rcube_imap(null);
            $all_recipients = array_merge(
                $IMAP->decode_address_list($headers['To'], null, true, $headers['charset']),
                $IMAP->decode_address_list($headers['Cc'], null, true, $headers['charset']),
                $IMAP->decode_address_list($headers['Bcc'], null, true, $headers['charset'])
            );
        } else {
            $all_recipients = array_merge(
                rcube_mime::decode_address_list($headers['To'], null, true, $headers['charset']),
                rcube_mime::decode_address_list($headers['Cc'], null, true, $headers['charset']),
                rcube_mime::decode_address_list($headers['Bcc'], null, true, $headers['charset'])
            );
        }

        require_once dirname(__FILE__) . '/automatic_addressbook_backend.php';
        $CONTACTS = new automatic_addressbook_backend($rcube->db, $rcube->user->ID);
    
        foreach ($all_recipients as $recipient) {
            // Bcc and Cc can be empty
            if ($recipient['mailto'] != '') {
                $contact = array(
                    'email' => $recipient['mailto'],
                    'name' => $recipient['name']
                    );

                // use email address part for name
                if (empty($contact['name']) || $contact['name'] == $contact['email']) {
                    $contact['name'] = ucfirst(preg_replace('/[\.\-]/', ' ',
                                               substr($contact['email'], 0, strpos($contact['email'], '@'))));
                }

                /* We only want to add the contact to the collected contacts
                 * address book if it is not already in an addressbook, so we
                 * first lookup in every address source.
                 */
                $book_types = (array)$rcube->config->get('autocomplete_addressbooks', 'sql');

                foreach ($book_types as $id) {
                    $abook = $rcube->get_address_book($id);
                    $previous_entries = $abook->search('email', $contact['email'], false, false);
      
                    if ($previous_entries->count) {
                        break;
                    }
                }
                if (!$previous_entries->count) {
                    $plugin = $rcube->plugins->exec_hook('contact_create', array('record' => $contact,
                                                                                  'source' => $this->abook_id));
                    if (!$plugin['abort']) {
                        $CONTACTS->insert($contact, false);
                    }
                }
            }
        }
    }
  

    /** 
     * Adds an address-book settings category in rc <= 0.5, not needed for rc >= 0.6
     *
     * @param array $args Hash array with list (hash array) of sections and  list (hash array) of sections
     * @return array Hash array with list (hash array) of sections and  list (hash array) of sections
     */
    public function add_addressbook_category($args)
    {
        $temp = $args['list']['server'];
        unset($args['list']['server']);
        $args['list']['addressbook']['id'] = 'addressbook';
        $args['list']['addressbook']['section'] = $this->gettext('addressbook');
        $args['list']['server'] = $temp;
        
        return $args;
    }


    /**
     * Adds a check-box to enable/disable automatic address collection.
     *
     * @param array $args Hash array containing section and preference blocks
     * @return array Hash array containing preference blocks with addressbook preferences
     */
    public function settings_table($args) 
    {
        if ($args['section'] == 'addressbook') {
            $use_auto_abook = rcube::get_instance()->config->get('use_auto_abook', true);
            $field_id = 'rcmfd_use_auto_abook';

            $checkbox = new html_checkbox(array(
                'name' => '_use_auto_abook', 
                'id' => $field_id, 'value' => 1
            ));
            $args['blocks']['automaticallycollected']['name'] = $this->gettext('automaticallycollected');
            $args['blocks']['automaticallycollected']['options']['use_subscriptions'] = array(
                'title' => html::label($field_id, rcube_utils::rep_specialchars_output($this->gettext('useautoabook'))),
                'content' => $checkbox->show($use_auto_abook ? 1 : 0),
            );

            $use_auto_abook_for_completion = rcube::get_instance()->config->get('use_auto_abook_for_completion', true);
            $field_id2 = 'rcmfd_use_auto_abook_for_completion';
            $checkbox2 = new html_checkbox(array(
                           'name' => '_use_auto_abook_for_completion',
                               'id' => $field_id2, 'value' => 1
                         ));
            $args['blocks']['automaticallycollected']['name'] = $this->gettext('automaticallycollected');
            $args['blocks']['automaticallycollected']['options']['use_autocompletion'] = array(
                'title' => html::label($field_id2, rcube_utils::rep_specialchars_output($this->gettext('useforcompletion'))),
                'content' => $checkbox2->show($use_auto_abook_for_completion ? 1 : 0),
            );
        }
        return $args;
    }

    /**
     * Save preferences
     *
     * @param array $args Hash array with prefs to be saved 
     * @return array $args Hash array with result: boolean, abort: boolean, prefs: array 
     */
    public function save_prefs($args) 
    {
        if ($args['section'] == 'addressbook') {
            $rcube = rcube::get_instance();
            $use_auto_abook = $rcube->config->get('use_auto_abook');
            $args['prefs']['use_auto_abook'] = isset($_POST['_use_auto_abook']) ? true : false;
            $use_auto_abook_for_completion = $rcube->config->get('use_auto_abook_for_completion');
            $args['prefs']['use_auto_abook_for_completion'] = isset($_POST['_use_auto_abook_for_completion']) ? true : false;
        }
        return $args;
    }

    /**
     * When a contact is added to a "regular" addressbook, take care to
     * delete it from collected addressbook if it was in.
     *
     * @param array $args Hash array with contact details
     * @return array record updated or abort
     */
    public function handle_doubles($args) 
    {
        $rcube = rcube::get_instance();
        if (!$rcube->config->get('use_auto_abook', true)) {
            return $args;
        }
        $moveto = $rcube->config->get('on_edit_move_to_default');

        if ($args['source'] == $this->abook_id && !empty($args['id']) && $moveto) {
            $args['source'] = $rcube->config->get('default_addressbook');
            $plugin = $rcube->plugins->exec_hook('contact_create', array('record' => $args['record'],
                                                                          'source' => $args['source']));
            if (!$plugin['abort']) {
                $CONTACTS = $rcube->get_address_book($args['source']);
                $insert_id = $CONTACTS->insert($args['record'], false);
            } else {
                $insert_id = $plugin['result'];
            }
            $rcube->output->show_message('automatic_addressbook.contactmoved', 'confirmation');
            $rcube->output->command('parent.list_contacts');
            return array('abort' => true, 'result' => $insert_id);
        }

        if ($args['source'] !== $this->abook_id) {
            foreach (array('email:home', 'email:work', 'email:other') as $email_field) {
                // Would trigger a warning with rc 0.5 without this if
                if ($args['record'][$email_field]) {
                    foreach ($args['record'][$email_field] as $email) {
                        $contact_emails[] = $email;
                    }
                }
            }
            // rc <= 0.5, retro-compatibility code, not needed for rc 0.6
            $contact_emails[] = $args['record']['email'];
            //

            foreach ($contact_emails as $contact_email) {
                if (!empty($contact_email)) {
                    $auto_abook = $rcube->get_address_book($this->abook_id);
                    $auto_abook->reset();
                    $collected_contact = $auto_abook->search('email', $contact_email, false, true);
                    while ($record = $collected_contact->iterate()) {
                        $plugin = $rcube->plugins->exec_hook('contact_delete', array('id' => $record['contact_id'],
                                                                                      'source' => $this->abook_id));
                        if (!$plugin['abort']) {
                            $auto_abook->delete($record['contact_id']);
                            $rcube->output->show_message('automatic_addressbook.contactremoved', 'confirmation');
                        }
                    }
                }
            }
            if ($rcube->task == "addressbook" &&
                $rcube->action == "copy" &&
                $_REQUEST['_source'] == $this->abook_id) {
                $rcube->output->command('parent.list_contacts');
            }
        }

        return $args;
    }
}
